// @eclipse-glsp/client/lib/sprotty-reexport.js
"use strict";
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/**
 * Sprotty still keeps the old class based action definitions as deprecated API.
 * This causes name clashes with the GLSP protocol actions when reexport sprotty.
 * The deprecated definitions are planned to be removed with Sprotty 1.0.0.
 * Until then we have to manually reexport the sprotty main index and exclude the clashing types.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectionResult = exports.SelectCommand = exports.SelectAllCommand = exports.GetSelectionCommand = exports.GetSelectionAction = exports.SetPopupModelCommand = exports.PopupHoverMouseListener = exports.HoverMouseListener = exports.HoverKeyListener = exports.HoverFeedbackCommand = exports.HoverFeedbackAction = exports.ClosePopupActionHandler = exports.AbstractHoverMouseListener = exports.SvgExporter = exports.RequestExportSvgAction = exports.ExportSvgPostprocessor = exports.ExportSvgKeyListener = exports.ExportSvgCommand = exports.SShapeElement = exports.layoutContainerFeature = exports.layoutableChildFeature = exports.isSizeable = exports.isLayoutContainer = exports.isLayoutableChild = exports.isBoundsAware = exports.isAlignable = exports.getAbsoluteClientBounds = exports.getAbsoluteBounds = exports.findChildrenAtPosition = exports.boundsFeature = exports.alignFeature = exports.SetBoundsCommand = exports.RequestBoundsCommand = exports.translatePoint = exports.translateBounds = exports.registerModelElement = exports.findParentByFeature = exports.findParent = exports.containsSome = exports.SParentElement = exports.SModelRoot = exports.SModelElement = exports.SChildElement = exports.ModelIndexImpl = exports.isParent = exports.createRandomId = exports.SetModelCommand = exports.SetBoundsAction = exports.CollapseExpandAllAction = exports.CollapseExpandAction = void 0;
exports.undoRedoModule = exports.selectModule = exports.routingModule = exports.openModule = exports.moveModule = exports.hoverModule = exports.fadeModule = exports.exportModule = exports.expandModule = exports.edgeLayoutModule = exports.edgeIntersectionModule = exports.decorationModule = exports.contextMenuModule = exports.commandPaletteModule = exports.buttonModule = exports.boundsModule = exports.graphModule = exports.defaultModule = exports.PointToPointLine = exports.intersection = exports.Diamond = exports.DiagramServerProxy = exports.ShapedPreRenderedElement = exports.RectangularPort = exports.RectangularNode = exports.PreRenderedElement = exports.HtmlRoot = exports.ForeignObjectElement = exports.DiamondNode = exports.CircularPort = exports.CircularNode = exports.SPort = exports.SNode = exports.SLabel = exports.SGraphIndex = exports.SGraph = exports.SEdge = exports.SCompartment = exports.ViewportResult = exports.ViewportAnimation = exports.SetViewportCommand = exports.GetViewportCommand = exports.GetViewportAction = exports.FitToScreenCommand = exports.CenterKeyboardListener = exports.CenterCommand = exports.BoundsAwareViewportCommand = exports.UpdateModelCommand = exports.SelectMouseListener = exports.SelectKeyboardListener = void 0;
exports.modelSourceModule = exports.zorderModule = exports.viewportModule = exports.updateModule = void 0;
__exportStar(require("@eclipse-glsp/protocol"), exports);
// sprotty-protocol
var actions_1 = require("sprotty-protocol/lib/actions");
Object.defineProperty(exports, "CollapseExpandAction", { enumerable: true, get: function () { return actions_1.CollapseExpandAction; } });
Object.defineProperty(exports, "CollapseExpandAllAction", { enumerable: true, get: function () { return actions_1.CollapseExpandAllAction; } });
Object.defineProperty(exports, "SetBoundsAction", { enumerable: true, get: function () { return actions_1.SetBoundsAction; } });
// ------------------ Base ------------------
__exportStar(require("sprotty/lib/base/actions/action-dispatcher"), exports);
__exportStar(require("sprotty/lib/base/actions/action-handler"), exports);
__exportStar(require("sprotty/lib/base/actions/diagram-locker"), exports);
__exportStar(require("sprotty/lib/base/animations/animation"), exports);
__exportStar(require("sprotty/lib/base/animations/animation-frame-syncer"), exports);
__exportStar(require("sprotty/lib/base/animations/easing"), exports);
__exportStar(require("sprotty/lib/base/commands/command"), exports);
__exportStar(require("sprotty/lib/base/commands/command-registration"), exports);
__exportStar(require("sprotty/lib/base/commands/command-stack"), exports);
__exportStar(require("sprotty/lib/base/commands/command-stack-options"), exports);
__exportStar(require("sprotty/lib/base/features/initialize-canvas"), exports);
var set_model_1 = require("sprotty/lib/base/features/set-model");
Object.defineProperty(exports, "SetModelCommand", { enumerable: true, get: function () { return set_model_1.SetModelCommand; } });
var smodel_1 = require("sprotty/lib/base/model/smodel");
Object.defineProperty(exports, "createRandomId", { enumerable: true, get: function () { return smodel_1.createRandomId; } });
Object.defineProperty(exports, "isParent", { enumerable: true, get: function () { return smodel_1.isParent; } });
Object.defineProperty(exports, "ModelIndexImpl", { enumerable: true, get: function () { return smodel_1.ModelIndexImpl; } });
Object.defineProperty(exports, "SChildElement", { enumerable: true, get: function () { return smodel_1.SChildElement; } });
Object.defineProperty(exports, "SModelElement", { enumerable: true, get: function () { return smodel_1.SModelElement; } });
Object.defineProperty(exports, "SModelRoot", { enumerable: true, get: function () { return smodel_1.SModelRoot; } });
Object.defineProperty(exports, "SParentElement", { enumerable: true, get: function () { return smodel_1.SParentElement; } });
__exportStar(require("sprotty/lib/base/model/smodel-extension"), exports);
__exportStar(require("sprotty/lib/base/model/smodel-factory"), exports);
var smodel_utils_1 = require("sprotty/lib/base/model/smodel-utils");
Object.defineProperty(exports, "containsSome", { enumerable: true, get: function () { return smodel_utils_1.containsSome; } });
Object.defineProperty(exports, "findParent", { enumerable: true, get: function () { return smodel_utils_1.findParent; } });
Object.defineProperty(exports, "findParentByFeature", { enumerable: true, get: function () { return smodel_utils_1.findParentByFeature; } });
Object.defineProperty(exports, "registerModelElement", { enumerable: true, get: function () { return smodel_utils_1.registerModelElement; } });
Object.defineProperty(exports, "translateBounds", { enumerable: true, get: function () { return smodel_utils_1.translateBounds; } });
Object.defineProperty(exports, "translatePoint", { enumerable: true, get: function () { return smodel_utils_1.translatePoint; } });
__exportStar(require("sprotty/lib/base/tool-manager/tool"), exports);
__exportStar(require("sprotty/lib/base/tool-manager/tool-manager"), exports);
__exportStar(require("sprotty/lib/base/ui-extensions/ui-extension"), exports);
__exportStar(require("sprotty/lib/base/ui-extensions/ui-extension-registry"), exports);
__exportStar(require("sprotty/lib/base/views/key-tool"), exports);
__exportStar(require("sprotty/lib/base/views/mouse-tool"), exports);
__exportStar(require("sprotty/lib/base/views/thunk-view"), exports);
__exportStar(require("sprotty/lib/base/views/view"), exports);
__exportStar(require("sprotty/lib/base/views/viewer"), exports);
__exportStar(require("sprotty/lib/base/views/viewer-cache"), exports);
__exportStar(require("sprotty/lib/base/views/viewer-options"), exports);
__exportStar(require("sprotty/lib/base/views/vnode-postprocessor"), exports);
__exportStar(require("sprotty/lib/base/views/vnode-utils"), exports);
// ------------------ Features ------------------
var bounds_manipulation_1 = require("sprotty/lib/features/bounds/bounds-manipulation");
Object.defineProperty(exports, "RequestBoundsCommand", { enumerable: true, get: function () { return bounds_manipulation_1.RequestBoundsCommand; } });
Object.defineProperty(exports, "SetBoundsCommand", { enumerable: true, get: function () { return bounds_manipulation_1.SetBoundsCommand; } });
__exportStar(require("sprotty/lib/features/bounds/hbox-layout"), exports);
__exportStar(require("sprotty/lib/features/bounds/hidden-bounds-updater"), exports);
__exportStar(require("sprotty/lib/features/bounds/layout"), exports);
var model_1 = require("sprotty/lib/features/bounds/model");
Object.defineProperty(exports, "alignFeature", { enumerable: true, get: function () { return model_1.alignFeature; } });
Object.defineProperty(exports, "boundsFeature", { enumerable: true, get: function () { return model_1.boundsFeature; } });
Object.defineProperty(exports, "findChildrenAtPosition", { enumerable: true, get: function () { return model_1.findChildrenAtPosition; } });
Object.defineProperty(exports, "getAbsoluteBounds", { enumerable: true, get: function () { return model_1.getAbsoluteBounds; } });
Object.defineProperty(exports, "getAbsoluteClientBounds", { enumerable: true, get: function () { return model_1.getAbsoluteClientBounds; } });
Object.defineProperty(exports, "isAlignable", { enumerable: true, get: function () { return model_1.isAlignable; } });
Object.defineProperty(exports, "isBoundsAware", { enumerable: true, get: function () { return model_1.isBoundsAware; } });
Object.defineProperty(exports, "isLayoutableChild", { enumerable: true, get: function () { return model_1.isLayoutableChild; } });
Object.defineProperty(exports, "isLayoutContainer", { enumerable: true, get: function () { return model_1.isLayoutContainer; } });
Object.defineProperty(exports, "isSizeable", { enumerable: true, get: function () { return model_1.isSizeable; } });
Object.defineProperty(exports, "layoutableChildFeature", { enumerable: true, get: function () { return model_1.layoutableChildFeature; } });
Object.defineProperty(exports, "layoutContainerFeature", { enumerable: true, get: function () { return model_1.layoutContainerFeature; } });
Object.defineProperty(exports, "SShapeElement", { enumerable: true, get: function () { return model_1.SShapeElement; } });
__exportStar(require("sprotty/lib/features/bounds/stack-layout"), exports);
__exportStar(require("sprotty/lib/features/bounds/vbox-layout"), exports);
__exportStar(require("sprotty/lib/features/bounds/views"), exports);
__exportStar(require("sprotty/lib/features/button/button-handler"), exports);
__exportStar(require("sprotty/lib/features/button/model"), exports);
__exportStar(require("sprotty/lib/features/command-palette/action-providers"), exports);
__exportStar(require("sprotty/lib/features/command-palette/command-palette"), exports);
__exportStar(require("sprotty/lib/features/context-menu/menu-providers"), exports);
__exportStar(require("sprotty/lib/features/context-menu/mouse-listener"), exports);
__exportStar(require("sprotty/lib/features/decoration/decoration-placer"), exports);
__exportStar(require("sprotty/lib/features/decoration/model"), exports);
__exportStar(require("sprotty/lib/features/decoration/views"), exports);
__exportStar(require("sprotty/lib/features/edge-intersection/intersection-finder"), exports);
__exportStar(require("sprotty/lib/features/edge-intersection/sweepline"), exports);
__exportStar(require("sprotty/lib/features/edge-layout/di.config"), exports);
__exportStar(require("sprotty/lib/features/edge-layout/edge-layout"), exports);
__exportStar(require("sprotty/lib/features/edge-layout/model"), exports);
__exportStar(require("sprotty/lib/features/edit/create"), exports);
__exportStar(require("sprotty/lib/features/edit/create-on-drag"), exports);
__exportStar(require("sprotty/lib/features/edit/delete"), exports);
__exportStar(require("sprotty/lib/features/edit/di.config"), exports);
__exportStar(require("sprotty/lib/features/edit/edit-label"), exports);
__exportStar(require("sprotty/lib/features/edit/edit-label-ui"), exports);
__exportStar(require("sprotty/lib/features/edit/edit-routing"), exports);
__exportStar(require("sprotty/lib/features/edit/model"), exports);
__exportStar(require("sprotty/lib/features/edit/reconnect"), exports);
__exportStar(require("sprotty/lib/features/expand/expand"), exports);
__exportStar(require("sprotty/lib/features/expand/model"), exports);
__exportStar(require("sprotty/lib/features/expand/views"), exports);
var export_1 = require("sprotty/lib/features/export/export");
Object.defineProperty(exports, "ExportSvgCommand", { enumerable: true, get: function () { return export_1.ExportSvgCommand; } });
Object.defineProperty(exports, "ExportSvgKeyListener", { enumerable: true, get: function () { return export_1.ExportSvgKeyListener; } });
Object.defineProperty(exports, "ExportSvgPostprocessor", { enumerable: true, get: function () { return export_1.ExportSvgPostprocessor; } });
Object.defineProperty(exports, "RequestExportSvgAction", { enumerable: true, get: function () { return export_1.RequestExportSvgAction; } });
__exportStar(require("sprotty/lib/features/export/model"), exports);
var svg_exporter_1 = require("sprotty/lib/features/export/svg-exporter");
Object.defineProperty(exports, "SvgExporter", { enumerable: true, get: function () { return svg_exporter_1.SvgExporter; } });
__exportStar(require("sprotty/lib/features/fade/fade"), exports);
__exportStar(require("sprotty/lib/features/fade/model"), exports);
var hover_1 = require("sprotty/lib/features/hover/hover");
Object.defineProperty(exports, "AbstractHoverMouseListener", { enumerable: true, get: function () { return hover_1.AbstractHoverMouseListener; } });
Object.defineProperty(exports, "ClosePopupActionHandler", { enumerable: true, get: function () { return hover_1.ClosePopupActionHandler; } });
Object.defineProperty(exports, "HoverFeedbackAction", { enumerable: true, get: function () { return hover_1.HoverFeedbackAction; } });
Object.defineProperty(exports, "HoverFeedbackCommand", { enumerable: true, get: function () { return hover_1.HoverFeedbackCommand; } });
Object.defineProperty(exports, "HoverKeyListener", { enumerable: true, get: function () { return hover_1.HoverKeyListener; } });
Object.defineProperty(exports, "HoverMouseListener", { enumerable: true, get: function () { return hover_1.HoverMouseListener; } });
Object.defineProperty(exports, "PopupHoverMouseListener", { enumerable: true, get: function () { return hover_1.PopupHoverMouseListener; } });
Object.defineProperty(exports, "SetPopupModelCommand", { enumerable: true, get: function () { return hover_1.SetPopupModelCommand; } });
__exportStar(require("sprotty/lib/features/hover/model"), exports);
__exportStar(require("sprotty/lib/features/move/model"), exports);
__exportStar(require("sprotty/lib/features/move/move"), exports);
__exportStar(require("sprotty/lib/features/move/snap"), exports);
__exportStar(require("sprotty/lib/features/nameable/model"), exports);
__exportStar(require("sprotty/lib/features/open/model"), exports);
__exportStar(require("sprotty/lib/features/open/open"), exports);
__exportStar(require("sprotty/lib/features/projection/model"), exports);
__exportStar(require("sprotty/lib/features/projection/views"), exports);
__exportStar(require("sprotty/lib/features/routing/abstract-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/anchor"), exports);
__exportStar(require("sprotty/lib/features/routing/bezier-anchors"), exports);
__exportStar(require("sprotty/lib/features/routing/bezier-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/manhattan-anchors"), exports);
__exportStar(require("sprotty/lib/features/routing/manhattan-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/model"), exports);
__exportStar(require("sprotty/lib/features/routing/polyline-anchors"), exports);
__exportStar(require("sprotty/lib/features/routing/polyline-edge-router"), exports);
__exportStar(require("sprotty/lib/features/routing/routing"), exports);
__exportStar(require("sprotty/lib/features/routing/views"), exports);
__exportStar(require("sprotty/lib/features/select/model"), exports);
var select_1 = require("sprotty/lib/features/select/select");
Object.defineProperty(exports, "GetSelectionAction", { enumerable: true, get: function () { return select_1.GetSelectionAction; } });
Object.defineProperty(exports, "GetSelectionCommand", { enumerable: true, get: function () { return select_1.GetSelectionCommand; } });
Object.defineProperty(exports, "SelectAllCommand", { enumerable: true, get: function () { return select_1.SelectAllCommand; } });
Object.defineProperty(exports, "SelectCommand", { enumerable: true, get: function () { return select_1.SelectCommand; } });
Object.defineProperty(exports, "SelectionResult", { enumerable: true, get: function () { return select_1.SelectionResult; } });
Object.defineProperty(exports, "SelectKeyboardListener", { enumerable: true, get: function () { return select_1.SelectKeyboardListener; } });
Object.defineProperty(exports, "SelectMouseListener", { enumerable: true, get: function () { return select_1.SelectMouseListener; } });
__exportStar(require("sprotty/lib/features/undo-redo/undo-redo"), exports);
__exportStar(require("sprotty/lib/features/update/model-matching"), exports);
var update_model_1 = require("sprotty/lib/features/update/update-model");
Object.defineProperty(exports, "UpdateModelCommand", { enumerable: true, get: function () { return update_model_1.UpdateModelCommand; } });
var center_fit_1 = require("sprotty/lib/features/viewport/center-fit");
Object.defineProperty(exports, "BoundsAwareViewportCommand", { enumerable: true, get: function () { return center_fit_1.BoundsAwareViewportCommand; } });
Object.defineProperty(exports, "CenterCommand", { enumerable: true, get: function () { return center_fit_1.CenterCommand; } });
Object.defineProperty(exports, "CenterKeyboardListener", { enumerable: true, get: function () { return center_fit_1.CenterKeyboardListener; } });
Object.defineProperty(exports, "FitToScreenCommand", { enumerable: true, get: function () { return center_fit_1.FitToScreenCommand; } });
__exportStar(require("sprotty/lib/features/viewport/model"), exports);
__exportStar(require("sprotty/lib/features/viewport/scroll"), exports);
var viewport_1 = require("sprotty/lib/features/viewport/viewport");
Object.defineProperty(exports, "GetViewportAction", { enumerable: true, get: function () { return viewport_1.GetViewportAction; } });
Object.defineProperty(exports, "GetViewportCommand", { enumerable: true, get: function () { return viewport_1.GetViewportCommand; } });
Object.defineProperty(exports, "SetViewportCommand", { enumerable: true, get: function () { return viewport_1.SetViewportCommand; } });
Object.defineProperty(exports, "ViewportAnimation", { enumerable: true, get: function () { return viewport_1.ViewportAnimation; } });
Object.defineProperty(exports, "ViewportResult", { enumerable: true, get: function () { return viewport_1.ViewportResult; } });
__exportStar(require("sprotty/lib/features/viewport/viewport-root"), exports);
__exportStar(require("sprotty/lib/features/viewport/zoom"), exports);
__exportStar(require("sprotty/lib/features/zorder/zorder"), exports);
var sgraph_1 = require("sprotty/lib/graph/sgraph");
Object.defineProperty(exports, "SCompartment", { enumerable: true, get: function () { return sgraph_1.SCompartment; } });
Object.defineProperty(exports, "SEdge", { enumerable: true, get: function () { return sgraph_1.SEdge; } });
Object.defineProperty(exports, "SGraph", { enumerable: true, get: function () { return sgraph_1.SGraph; } });
Object.defineProperty(exports, "SGraphIndex", { enumerable: true, get: function () { return sgraph_1.SGraphIndex; } });
Object.defineProperty(exports, "SLabel", { enumerable: true, get: function () { return sgraph_1.SLabel; } });
Object.defineProperty(exports, "SNode", { enumerable: true, get: function () { return sgraph_1.SNode; } });
Object.defineProperty(exports, "SPort", { enumerable: true, get: function () { return sgraph_1.SPort; } });
// ------------------ Graph ------------------
__exportStar(require("sprotty/lib/graph/sgraph-factory"), exports);
__exportStar(require("sprotty/lib/graph/views"), exports);
__exportStar(require("sprotty/lib/lib/generic-views"), exports);
__exportStar(require("sprotty/lib/lib/html-views"), exports);
__exportStar(require("sprotty/lib/lib/jsx"), exports);
var model_2 = require("sprotty/lib/lib/model");
Object.defineProperty(exports, "CircularNode", { enumerable: true, get: function () { return model_2.CircularNode; } });
Object.defineProperty(exports, "CircularPort", { enumerable: true, get: function () { return model_2.CircularPort; } });
Object.defineProperty(exports, "DiamondNode", { enumerable: true, get: function () { return model_2.DiamondNode; } });
Object.defineProperty(exports, "ForeignObjectElement", { enumerable: true, get: function () { return model_2.ForeignObjectElement; } });
Object.defineProperty(exports, "HtmlRoot", { enumerable: true, get: function () { return model_2.HtmlRoot; } });
Object.defineProperty(exports, "PreRenderedElement", { enumerable: true, get: function () { return model_2.PreRenderedElement; } });
Object.defineProperty(exports, "RectangularNode", { enumerable: true, get: function () { return model_2.RectangularNode; } });
Object.defineProperty(exports, "RectangularPort", { enumerable: true, get: function () { return model_2.RectangularPort; } });
Object.defineProperty(exports, "ShapedPreRenderedElement", { enumerable: true, get: function () { return model_2.ShapedPreRenderedElement; } });
// ------------------ Library ------------------
__exportStar(require("sprotty/lib/lib/modules"), exports);
__exportStar(require("sprotty/lib/lib/svg-views"), exports);
// ------------------ Model Source ------------------
__exportStar(require("sprotty/lib/model-source/commit-model"), exports);
var diagram_server_1 = require("sprotty/lib/model-source/diagram-server");
Object.defineProperty(exports, "DiagramServerProxy", { enumerable: true, get: function () { return diagram_server_1.DiagramServerProxy; } });
__exportStar(require("sprotty/lib/model-source/local-model-source"), exports);
__exportStar(require("sprotty/lib/model-source/logging"), exports);
__exportStar(require("sprotty/lib/model-source/model-source"), exports);
__exportStar(require("sprotty/lib/model-source/websocket"), exports);
// ------------------ Utilities ------------------
__exportStar(require("sprotty/lib/utils/browser"), exports);
__exportStar(require("sprotty/lib/utils/codicon"), exports);
__exportStar(require("sprotty/lib/utils/color"), exports);
var geometry_1 = require("sprotty/lib/utils/geometry");
Object.defineProperty(exports, "Diamond", { enumerable: true, get: function () { return geometry_1.Diamond; } });
Object.defineProperty(exports, "intersection", { enumerable: true, get: function () { return geometry_1.intersection; } });
Object.defineProperty(exports, "PointToPointLine", { enumerable: true, get: function () { return geometry_1.PointToPointLine; } });
__exportStar(require("sprotty/lib/utils/inversify"), exports);
__exportStar(require("sprotty/lib/utils/logging"), exports);
__exportStar(require("sprotty/lib/utils/registry"), exports);
const di_config_1 = require("sprotty/lib/base/di.config");
exports.defaultModule = di_config_1.default;
const di_config_2 = require("sprotty/lib/features/bounds/di.config");
exports.boundsModule = di_config_2.default;
const di_config_3 = require("sprotty/lib/features/button/di.config");
exports.buttonModule = di_config_3.default;
const di_config_4 = require("sprotty/lib/features/command-palette/di.config");
exports.commandPaletteModule = di_config_4.default;
const di_config_5 = require("sprotty/lib/features/context-menu/di.config");
exports.contextMenuModule = di_config_5.default;
const di_config_6 = require("sprotty/lib/features/decoration/di.config");
exports.decorationModule = di_config_6.default;
const di_config_7 = require("sprotty/lib/features/edge-intersection/di.config");
exports.edgeIntersectionModule = di_config_7.default;
const di_config_8 = require("sprotty/lib/features/edge-layout/di.config");
exports.edgeLayoutModule = di_config_8.default;
const di_config_9 = require("sprotty/lib/features/expand/di.config");
exports.expandModule = di_config_9.default;
const di_config_10 = require("sprotty/lib/features/export/di.config");
exports.exportModule = di_config_10.default;
const di_config_11 = require("sprotty/lib/features/fade/di.config");
exports.fadeModule = di_config_11.default;
const di_config_12 = require("sprotty/lib/features/hover/di.config");
exports.hoverModule = di_config_12.default;
const di_config_13 = require("sprotty/lib/features/move/di.config");
exports.moveModule = di_config_13.default;
const di_config_14 = require("sprotty/lib/features/open/di.config");
exports.openModule = di_config_14.default;
const di_config_15 = require("sprotty/lib/features/routing/di.config");
exports.routingModule = di_config_15.default;
const di_config_16 = require("sprotty/lib/features/select/di.config");
exports.selectModule = di_config_16.default;
const di_config_17 = require("sprotty/lib/features/undo-redo/di.config");
exports.undoRedoModule = di_config_17.default;
const di_config_18 = require("sprotty/lib/features/update/di.config");
exports.updateModule = di_config_18.default;
const di_config_19 = require("sprotty/lib/features/viewport/di.config");
exports.viewportModule = di_config_19.default;
const di_config_20 = require("sprotty/lib/features/zorder/di.config");
exports.zorderModule = di_config_20.default;
const di_config_21 = require("sprotty/lib/graph/di.config");
exports.graphModule = di_config_21.default;
const di_config_22 = require("sprotty/lib/model-source/di.config");
exports.modelSourceModule = di_config_22.default;
