// @eclipse-glsp/protocol/lib/action-protocol/clipboard.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const clipboard_1 = require("./clipboard");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Clipboard actions', () => {
    describe('RequestClipboardDataAction', () => {
        const requestClipboardDataAction = {
            kind: 'requestClipboardData',
            requestId: '',
            editorContext: { selectedElementIds: [] }
        };
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                (0, chai_1.expect)(clipboard_1.RequestClipboardDataAction.is(requestClipboardDataAction)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(clipboard_1.RequestClipboardDataAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(clipboard_1.RequestClipboardDataAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: clipboard_1.RequestClipboardDataAction.KIND,
                    editorContext: { selectedElementIds: ['element1'] },
                    requestId: ''
                };
                const { editorContext } = expected;
                (0, chai_1.expect)(clipboard_1.RequestClipboardDataAction.create(editorContext)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: clipboard_1.RequestClipboardDataAction.KIND,
                    editorContext: { selectedElementIds: ['element1'] },
                    requestId: '100'
                };
                const { editorContext, requestId } = expected;
                (0, chai_1.expect)(clipboard_1.RequestClipboardDataAction.create(editorContext, { requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('SetClipboardDataAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    clipboardData: { format: '' },
                    kind: 'setClipboardData',
                    responseId: ''
                };
                (0, chai_1.expect)(clipboard_1.SetClipboardDataAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(clipboard_1.SetClipboardDataAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(clipboard_1.SetClipboardDataAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    clipboardData: { format: 'json' },
                    kind: 'setClipboardData',
                    responseId: ''
                };
                const { clipboardData } = expected;
                (0, chai_1.expect)(clipboard_1.SetClipboardDataAction.create(clipboardData)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    clipboardData: { format: 'json' },
                    kind: 'setClipboardData',
                    responseId: '600'
                };
                const { clipboardData, responseId } = expected;
                (0, chai_1.expect)(clipboard_1.SetClipboardDataAction.create(clipboardData, { responseId })).to.deep.equals(expected);
            });
        });
    });
    describe('CutOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'cut',
                    isOperation: true,
                    editorContext: { selectedElementIds: [] }
                };
                (0, chai_1.expect)(clipboard_1.CutOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(clipboard_1.CutOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(clipboard_1.CutOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'cut',
                    isOperation: true,
                    editorContext: { selectedElementIds: ['element1'] }
                };
                const { editorContext } = expected;
                (0, chai_1.expect)(clipboard_1.CutOperation.create(editorContext)).to.deep.equals(expected);
            });
        });
    });
    describe('PasteOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'paste',
                    isOperation: true,
                    editorContext: { selectedElementIds: [] },
                    clipboardData: { format: '' }
                };
                (0, chai_1.expect)(clipboard_1.PasteOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(clipboard_1.PasteOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(clipboard_1.PasteOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'paste',
                    isOperation: true,
                    clipboardData: { format: 'string' },
                    editorContext: { selectedElementIds: ['element1'] }
                };
                const { editorContext, clipboardData } = expected;
                (0, chai_1.expect)(clipboard_1.PasteOperation.create({ clipboardData, editorContext })).to.deep.equals(expected);
            });
        });
    });
});
