// @eclipse-glsp/protocol/lib/action-protocol/element-navigation.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const element_navigation_1 = require("./element-navigation");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Element navigation actions', () => {
    describe('NavigationTarget', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const target = {
                    uri: ''
                };
                (0, chai_1.expect)(element_navigation_1.NavigationTarget.is(target)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_navigation_1.NavigationTarget.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_navigation_1.NavigationTarget.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('hasArguments', () => {
            it('should return true for an object having the correct type and a value for all required interface properties and a defined `args` property', () => {
                const target = { uri: 'myuri', args: { some: 'args' } };
                (0, chai_1.expect)(element_navigation_1.NavigationTarget.hasArguments(target)).to.be.true;
            });
            it('should return false for an object having the correct type and a value for all required interface properties and an undefined `args property', () => {
                const target = { uri: 'myuri' };
                (0, chai_1.expect)(element_navigation_1.NavigationTarget.hasArguments(target)).to.be.false;
            });
        });
        describe('addArgument', () => {
            it('should assign a new `args` property to the given target and add the given key-value pair to the args object', () => {
                var _a;
                const target = { uri: 'myuri' };
                element_navigation_1.NavigationTarget.addArgument(target, 'some', 'argument');
                (0, chai_1.expect)((_a = target.args) === null || _a === void 0 ? void 0 : _a.some).to.be.equal('argument');
            });
        });
        describe('getElementIds', () => {
            it('should return the value for the ELEMENT_IDS key of the args object of the given target as a string array', () => {
                const id1 = 'someId';
                const id2 = 'anotherId';
                const target = { uri: 'myuri', args: { [element_navigation_1.NavigationTarget.ELEMENT_IDS]: `${id1}&${id2}` } };
                const elementIds = element_navigation_1.NavigationTarget.getElementIds(target);
                (0, chai_1.expect)(elementIds.length).to.be.equal(2);
                (0, chai_1.expect)(elementIds).to.include(id1);
                (0, chai_1.expect)(elementIds).to.include(id2);
            });
            it('should return an empty array for the given target with an undefined args property', () => {
                const target = { uri: 'myuri' };
                const elementIds = element_navigation_1.NavigationTarget.getElementIds(target);
                (0, chai_1.expect)(elementIds.length).to.be.equal(0);
            });
        });
        describe('setElementIds', () => {
            it('(should store the given element id array as a joined string value for the ELEMENT_IDS key', () => {
                var _a;
                const target = { uri: 'myuri' };
                const id1 = 'someId';
                const id2 = 'anotherId';
                element_navigation_1.NavigationTarget.setElementIds(target, id1, id2);
                (0, chai_1.expect)((_a = target.args) === null || _a === void 0 ? void 0 : _a[element_navigation_1.NavigationTarget.ELEMENT_IDS]).to.be.equal(`${id1}&${id2}`);
            });
        });
        describe('setTextPosition', () => {
            it('(should store the given text position in the args property of the given target with separate keys for column and line.', () => {
                var _a, _b;
                const target = { uri: 'myuri' };
                const pos = { character: 5, line: 2 };
                const { character: column, line } = pos;
                element_navigation_1.NavigationTarget.setTextPosition(target, pos);
                (0, chai_1.expect)((_a = target.args) === null || _a === void 0 ? void 0 : _a[element_navigation_1.NavigationTarget.TEXT_COLUMN]).to.be.equal(column);
                (0, chai_1.expect)((_b = target.args) === null || _b === void 0 ? void 0 : _b[element_navigation_1.NavigationTarget.TEXT_LINE]).to.be.equal(line);
            });
        });
        describe('getTextPosition', () => {
            it('should return the text position composed of its correspond keys in the args object of the given target', () => {
                const position = {
                    character: 5,
                    line: 2
                };
                const target = {
                    uri: 'myuri',
                    args: { [element_navigation_1.NavigationTarget.TEXT_COLUMN]: position.character, [element_navigation_1.NavigationTarget.TEXT_LINE]: position.line }
                };
                const result = element_navigation_1.NavigationTarget.getTextPosition(target);
                (0, chai_1.expect)(result).to.be.not.undefined;
                (0, chai_1.expect)(result).to.be.deep.equal(position);
            });
            it('should return undefined for the given target with an undefined args property', () => {
                const target = { uri: 'myuri' };
                const result = element_navigation_1.NavigationTarget.getTextPosition(target);
                (0, chai_1.expect)(result).to.be.undefined;
            });
        });
    });
    describe('RequestNavigationTargetsAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'requestNavigationTargets',
                    requestId: '',
                    editorContext: { selectedElementIds: [] },
                    targetTypeId: ''
                };
                (0, chai_1.expect)(element_navigation_1.RequestNavigationTargetsAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_navigation_1.RequestNavigationTargetsAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_navigation_1.RequestNavigationTargetsAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'requestNavigationTargets',
                    requestId: '',
                    editorContext: { selectedElementIds: ['selected'] },
                    targetTypeId: 'someId'
                };
                const { editorContext, targetTypeId } = expected;
                (0, chai_1.expect)(element_navigation_1.RequestNavigationTargetsAction.create({ editorContext, targetTypeId })).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'requestNavigationTargets',
                    requestId: 'myRequest',
                    editorContext: { selectedElementIds: ['selected'] },
                    targetTypeId: 'someId'
                };
                const { editorContext, targetTypeId, requestId } = expected;
                (0, chai_1.expect)(element_navigation_1.RequestNavigationTargetsAction.create({ editorContext, targetTypeId, requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('SetNavigationTargetsAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'setNavigationTargets',
                    responseId: '',
                    targets: []
                };
                (0, chai_1.expect)(element_navigation_1.SetNavigationTargetsAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_navigation_1.SetNavigationTargetsAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_navigation_1.SetNavigationTargetsAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'setNavigationTargets',
                    responseId: '',
                    targets: [{ uri: 'someUri' }]
                };
                const { targets } = expected;
                (0, chai_1.expect)(element_navigation_1.SetNavigationTargetsAction.create(targets)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'setNavigationTargets',
                    responseId: 'myResponse',
                    targets: [{ uri: 'someUri' }]
                };
                const { targets, responseId } = expected;
                (0, chai_1.expect)(element_navigation_1.SetNavigationTargetsAction.create(targets, { responseId })).to.deep.equals(expected);
            });
        });
    });
    describe('NavigateToTargetAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'navigateToTarget',
                    target: { uri: '' }
                };
                (0, chai_1.expect)(element_navigation_1.NavigateToTargetAction.is(action)).to.be.true;
            });
        });
        it('should return false for `undefined`', () => {
            (0, chai_1.expect)(element_navigation_1.NavigateToTargetAction.is(undefined)).to.be.false;
        });
        it('should return false for an object that does not have all required interface properties', () => {
            (0, chai_1.expect)(element_navigation_1.NavigateToTargetAction.is({ kind: 'notTheRightOne' })).to.be.false;
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'navigateToTarget',
                    target: { uri: 'myUri' }
                };
                const { target } = expected;
                (0, chai_1.expect)(element_navigation_1.NavigateToTargetAction.create(target)).to.deep.equals(expected);
            });
        });
    });
    describe('ResolveNavigationTargetAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'resolveNavigationTarget',
                    requestId: '',
                    navigationTarget: { uri: '' }
                };
                (0, chai_1.expect)(element_navigation_1.ResolveNavigationTargetAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_navigation_1.ResolveNavigationTargetAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_navigation_1.ResolveNavigationTargetAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'resolveNavigationTarget',
                    requestId: '',
                    navigationTarget: { uri: '' }
                };
                const { navigationTarget } = expected;
                (0, chai_1.expect)(element_navigation_1.ResolveNavigationTargetAction.create(navigationTarget)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'resolveNavigationTarget',
                    requestId: 'myRequest',
                    navigationTarget: { uri: '' }
                };
                const { navigationTarget, requestId } = expected;
                (0, chai_1.expect)(element_navigation_1.ResolveNavigationTargetAction.create(navigationTarget, { requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('SetResolvedNavigationTargetAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'setResolvedNavigationTarget',
                    responseId: '',
                    elementIds: ['']
                };
                (0, chai_1.expect)(element_navigation_1.SetResolvedNavigationTargetAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_navigation_1.SetResolvedNavigationTargetAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_navigation_1.SetResolvedNavigationTargetAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'setResolvedNavigationTarget',
                    responseId: '',
                    elementIds: ['']
                };
                const { elementIds } = expected;
                (0, chai_1.expect)(element_navigation_1.SetResolvedNavigationTargetAction.create(elementIds)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'setResolvedNavigationTarget',
                    responseId: '',
                    elementIds: ['']
                };
                const { elementIds, responseId } = expected;
                (0, chai_1.expect)(element_navigation_1.SetResolvedNavigationTargetAction.create(elementIds, { responseId })).to.deep.equals(expected);
            });
        });
    });
    describe('NavigateToExternalTargetAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'navigateToExternalTarget',
                    target: { uri: '' }
                };
                (0, chai_1.expect)(element_navigation_1.NavigateToExternalTargetAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_navigation_1.NavigateToExternalTargetAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_navigation_1.NavigateToExternalTargetAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'navigateToExternalTarget',
                    target: { uri: '' }
                };
                const { target } = expected;
                (0, chai_1.expect)(element_navigation_1.NavigateToExternalTargetAction.create(target)).to.deep.equals(expected);
            });
        });
    });
});
