// @eclipse-glsp/protocol/lib/action-protocol/element-text-editing.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const element_text_editing_1 = require("./element-text-editing");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Element text editing actions', () => {
    describe('RequestEditValidationAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'requestEditValidation',
                    requestId: '',
                    contextId: '',
                    modelElementId: '',
                    text: ''
                };
                (0, chai_1.expect)(element_text_editing_1.RequestEditValidationAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_text_editing_1.RequestEditValidationAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_text_editing_1.RequestEditValidationAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'requestEditValidation',
                    requestId: '',
                    contextId: 'myContext',
                    modelElementId: 'myModelElement',
                    text: 'someText'
                };
                const { contextId, modelElementId, text } = expected;
                (0, chai_1.expect)(element_text_editing_1.RequestEditValidationAction.create({ contextId, modelElementId, text })).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'requestEditValidation',
                    requestId: 'myRequest',
                    contextId: 'myContext',
                    modelElementId: 'myModelElement',
                    text: 'someText'
                };
                const { contextId, modelElementId, text, requestId } = expected;
                (0, chai_1.expect)(element_text_editing_1.RequestEditValidationAction.create({ contextId, modelElementId, text, requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('SetEditValidationResultAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'setEditValidationResult',
                    responseId: '',
                    status: { severity: element_text_editing_1.ValidationStatus.Severity.OK }
                };
                (0, chai_1.expect)(element_text_editing_1.SetEditValidationResultAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_text_editing_1.SetEditValidationResultAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_text_editing_1.SetEditValidationResultAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'setEditValidationResult',
                    responseId: '',
                    status: { severity: element_text_editing_1.ValidationStatus.Severity.OK }
                };
                const { status } = expected;
                (0, chai_1.expect)(element_text_editing_1.SetEditValidationResultAction.create(status)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'setEditValidationResult',
                    responseId: 'myRespsone',
                    args: { some: 'args' },
                    status: { severity: element_text_editing_1.ValidationStatus.Severity.OK }
                };
                const { status, args, responseId } = expected;
                (0, chai_1.expect)(element_text_editing_1.SetEditValidationResultAction.create(status, { args, responseId })).to.deep.equals(expected);
            });
        });
    });
    describe('ApplyLabelEditOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'applyLabelEdit',
                    isOperation: true,
                    labelId: '',
                    text: ''
                };
                (0, chai_1.expect)(element_text_editing_1.ApplyLabelEditOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_text_editing_1.ApplyLabelEditOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_text_editing_1.ApplyLabelEditOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'applyLabelEdit',
                    isOperation: true,
                    labelId: 'myLabel',
                    text: 'myText'
                };
                const { labelId, text } = expected;
                (0, chai_1.expect)(element_text_editing_1.ApplyLabelEditOperation.create({ labelId, text })).to.deep.equals(expected);
            });
        });
    });
});
