// @eclipse-glsp/protocol/lib/action-protocol/element-creation.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const sprotty_protocol_1 = require("sprotty-protocol");
const element_creation_1 = require("./element-creation");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Element creation operations', () => {
    describe('CreateOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'CreateEdgeOperation',
                    isOperation: true,
                    elementTypeId: 'someType'
                };
                (0, chai_1.expect)(element_creation_1.CreateOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_creation_1.CreateOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_creation_1.CreateOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
    });
    describe('CreateNodeOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'createNode',
                    isOperation: true,
                    elementTypeId: ''
                };
                (0, chai_1.expect)(element_creation_1.CreateNodeOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_creation_1.CreateNodeOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_creation_1.CreateNodeOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'createNode',
                    isOperation: true,
                    elementTypeId: 'someNode'
                };
                const { elementTypeId } = expected;
                (0, chai_1.expect)(element_creation_1.CreateNodeOperation.create(elementTypeId)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'createNode',
                    elementTypeId: 'elementType',
                    containerId: 'container',
                    location: sprotty_protocol_1.Point.ORIGIN,
                    isOperation: true,
                    args: { some: 'args' }
                };
                const { elementTypeId, containerId, location, args } = expected;
                (0, chai_1.expect)(element_creation_1.CreateNodeOperation.create(elementTypeId, { args, location, containerId })).to.deep.equals(expected);
            });
        });
    });
    describe('CreateEdgeOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'createEdge',
                    isOperation: true,
                    elementTypeId: '',
                    sourceElementId: '',
                    targetElementId: ''
                };
                (0, chai_1.expect)(element_creation_1.CreateEdgeOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_creation_1.CreateEdgeOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_creation_1.CreateEdgeOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'createEdge',
                    isOperation: true,
                    elementTypeId: 'someNode',
                    sourceElementId: 'source',
                    targetElementId: 'target'
                };
                const { elementTypeId, sourceElementId, targetElementId } = expected;
                (0, chai_1.expect)(element_creation_1.CreateEdgeOperation.create({ elementTypeId, sourceElementId, targetElementId })).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'createEdge',
                    isOperation: true,
                    elementTypeId: 'someNode',
                    sourceElementId: 'source',
                    targetElementId: 'target',
                    args: { some: 'args' }
                };
                const { elementTypeId, sourceElementId, targetElementId, args } = expected;
                (0, chai_1.expect)(element_creation_1.CreateEdgeOperation.create({ elementTypeId, sourceElementId, targetElementId, args })).to.deep.equals(expected);
            });
        });
    });
    describe('DeleteElementOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const operation = {
                    kind: 'deleteElement',
                    isOperation: true,
                    elementIds: []
                };
                (0, chai_1.expect)(element_creation_1.DeleteElementOperation.is(operation)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(element_creation_1.DeleteElementOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(element_creation_1.DeleteElementOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'deleteElement',
                    isOperation: true,
                    elementIds: ['deleteMe']
                };
                const { elementIds } = expected;
                (0, chai_1.expect)(element_creation_1.DeleteElementOperation.create(elementIds)).to.deep.equals(expected);
            });
        });
    });
});
