// @eclipse-glsp/protocol/lib/action-protocol/model-data.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const model_data_1 = require("./model-data");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Model data actions', () => {
    describe('RequestModelAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                it('RequestModelAction.is with valid action type', () => {
                    const action = {
                        kind: 'requestModel',
                        requestId: ''
                    };
                    (0, chai_1.expect)(model_data_1.RequestModelAction.is(action)).to.be.true;
                });
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_data_1.RequestModelAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_data_1.RequestModelAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'requestModel',
                    requestId: ''
                };
                (0, chai_1.expect)(model_data_1.RequestModelAction.create()).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'requestModel',
                    requestId: 'myRequestId',
                    options: { some: 'option' }
                };
                const { requestId, options } = expected;
                (0, chai_1.expect)(model_data_1.RequestModelAction.create({ options, requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('SetModelAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'setModel',
                    responseId: '',
                    newRoot: { id: '', type: '' }
                };
                (0, chai_1.expect)(model_data_1.SetModelAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_data_1.SetModelAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_data_1.SetModelAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'setModel',
                    responseId: '',
                    newRoot: { id: 'myId', type: 'myType' }
                };
                const { newRoot } = expected;
                (0, chai_1.expect)(model_data_1.SetModelAction.create(newRoot)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'setModel',
                    responseId: 'myResponse',
                    newRoot: { id: '', type: '' }
                };
                const { newRoot, responseId } = expected;
                (0, chai_1.expect)(model_data_1.SetModelAction.create(newRoot, { responseId })).to.deep.equals(expected);
            });
        });
    });
    describe('UpdateModelAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'updateModel',
                    newRoot: { id: '', type: '' }
                };
                (0, chai_1.expect)(model_data_1.UpdateModelAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_data_1.UpdateModelAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_data_1.UpdateModelAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'updateModel',
                    animate: true,
                    newRoot: { id: 'myId', type: 'myType' }
                };
                const { newRoot } = expected;
                (0, chai_1.expect)(model_data_1.UpdateModelAction.create(newRoot)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'updateModel',
                    newRoot: { id: 'myId', type: 'myType' },
                    animate: false
                };
                const { newRoot, animate } = expected;
                (0, chai_1.expect)(model_data_1.UpdateModelAction.create(newRoot, { animate })).to.deep.equals(expected);
            });
        });
    });
    describe('ModelSourceChangedAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'modelSourceChanged',
                    modelSourceName: ''
                };
                (0, chai_1.expect)(model_data_1.ModelSourceChangedAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_data_1.ModelSourceChangedAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_data_1.ModelSourceChangedAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'modelSourceChanged',
                    modelSourceName: 'myModelSource'
                };
                const { modelSourceName } = expected;
                (0, chai_1.expect)(model_data_1.ModelSourceChangedAction.create(modelSourceName)).to.deep.equals(expected);
            });
        });
    });
});
