// @eclipse-glsp/protocol/lib/action-protocol/model-saving.spec.js
"use strict";
/********************************************************************************
 * Copyright (c) 2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const model_saving_1 = require("./model-saving");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Model saving actions', () => {
    describe('SaveModelAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'saveModel'
                };
                (0, chai_1.expect)(model_saving_1.SaveModelAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_saving_1.SaveModelAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_saving_1.SaveModelAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'saveModel'
                };
                (0, chai_1.expect)(model_saving_1.SaveModelAction.create()).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'saveModel',
                    fileUri: 'myUri'
                };
                const { fileUri } = expected;
                (0, chai_1.expect)(model_saving_1.SaveModelAction.create({ fileUri })).to.deep.equals(expected);
            });
        });
    });
    describe('SetDirtyStateAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'setDirtyState',
                    isDirty: true
                };
                (0, chai_1.expect)(model_saving_1.SetDirtyStateAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_saving_1.SetDirtyStateAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_saving_1.SetDirtyStateAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'setDirtyState',
                    isDirty: true
                };
                const { isDirty } = expected;
                (0, chai_1.expect)(model_saving_1.SetDirtyStateAction.create(isDirty)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'setDirtyState',
                    isDirty: false,
                    reason: 'save'
                };
                const { isDirty, reason } = expected;
                (0, chai_1.expect)(model_saving_1.SetDirtyStateAction.create(isDirty, { reason })).to.deep.equals(expected);
            });
        });
    });
    describe('RequestExportSvgAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'requestExportSvg',
                    requestId: ''
                };
                (0, chai_1.expect)(model_saving_1.RequestExportSvgAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_saving_1.RequestExportSvgAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_saving_1.RequestExportSvgAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'requestExportSvg',
                    requestId: ''
                };
                (0, chai_1.expect)(model_saving_1.RequestExportSvgAction.create()).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'requestExportSvg',
                    requestId: 'myRequest'
                };
                const { requestId } = expected;
                (0, chai_1.expect)(model_saving_1.RequestExportSvgAction.create({ requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('ExportSvgAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'exportSvg',
                    responseId: '',
                    svg: ''
                };
                (0, chai_1.expect)(model_saving_1.ExportSvgAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_saving_1.ExportSvgAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_saving_1.ExportSvgAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'exportSvg',
                    responseId: '',
                    svg: 'someSvg'
                };
                const { svg } = expected;
                (0, chai_1.expect)(model_saving_1.ExportSvgAction.create(svg)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'exportSvg',
                    responseId: 'responseId',
                    svg: 'someSvg'
                };
                const { svg, responseId } = expected;
                (0, chai_1.expect)(model_saving_1.ExportSvgAction.create(svg, { responseId })).to.deep.equals(expected);
            });
        });
    });
});
