// @eclipse-glsp/protocol/lib/action-protocol/model-layout.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const sprotty_protocol_1 = require("sprotty-protocol");
const model_layout_1 = require("./model-layout");
/**
 * Tests for the utility functions declared in the namespaces of the protocol
 * action definitions.
 */
describe('Model layout actions', () => {
    describe('RequestBoundsAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'requestBounds',
                    requestId: '',
                    newRoot: { id: '', type: '' }
                };
                (0, chai_1.expect)(model_layout_1.RequestBoundsAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_layout_1.RequestBoundsAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_layout_1.RequestBoundsAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'requestBounds',
                    requestId: '',
                    newRoot: { id: 'myId', type: 'myType' }
                };
                const { newRoot } = expected;
                (0, chai_1.expect)(model_layout_1.RequestBoundsAction.create(newRoot)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'requestBounds',
                    requestId: 'myRequest',
                    newRoot: { id: 'myId', type: 'myType' }
                };
                const { newRoot, requestId } = expected;
                (0, chai_1.expect)(model_layout_1.RequestBoundsAction.create(newRoot, { requestId })).to.deep.equals(expected);
            });
        });
    });
    describe('ComputedBoundsAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'computedBounds',
                    responseId: '',
                    bounds: []
                };
                (0, chai_1.expect)(model_layout_1.ComputedBoundsAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_layout_1.ComputedBoundsAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_layout_1.ComputedBoundsAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'computedBounds',
                    responseId: '',
                    bounds: [{ elementId: '', newSize: sprotty_protocol_1.Dimension.EMPTY, newPosition: sprotty_protocol_1.Point.ORIGIN }]
                };
                const { bounds } = expected;
                (0, chai_1.expect)(model_layout_1.ComputedBoundsAction.create(bounds)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'computedBounds',
                    responseId: 'myResponse',
                    alignments: [{ elementId: 'myElement', newAlignment: sprotty_protocol_1.Point.ORIGIN }],
                    revision: 5,
                    bounds: [{ elementId: '', newSize: sprotty_protocol_1.Dimension.EMPTY, newPosition: sprotty_protocol_1.Point.ORIGIN }]
                };
                const { bounds, responseId, alignments, revision } = expected;
                (0, chai_1.expect)(model_layout_1.ComputedBoundsAction.create(bounds, { responseId, alignments, revision })).to.deep.equals(expected);
            });
        });
    });
    describe('LayoutOperation', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'layout',
                    isOperation: true,
                    elementIds: []
                };
                (0, chai_1.expect)(model_layout_1.LayoutOperation.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(model_layout_1.LayoutOperation.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(model_layout_1.LayoutOperation.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments', () => {
                const expected = {
                    kind: 'layout',
                    isOperation: true,
                    elementIds: ['myElements']
                };
                const { elementIds } = expected;
                (0, chai_1.expect)(model_layout_1.LayoutOperation.create(elementIds)).to.deep.equals(expected);
            });
        });
    });
});
