// @eclipse-glsp/protocol/lib/action-protocol/model-saving.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExportSvgAction = exports.RequestExportSvgAction = exports.SetDirtyStateAction = exports.SaveModelAction = void 0;
/********************************************************************************
 * Copyright (c) 2021-2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const type_util_1 = require("../utils/type-util");
const base_protocol_1 = require("./base-protocol");
var SaveModelAction;
(function (SaveModelAction) {
    SaveModelAction.KIND = 'saveModel';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, SaveModelAction.KIND);
    }
    SaveModelAction.is = is;
    function create(options = {}) {
        return Object.assign({ kind: SaveModelAction.KIND }, options);
    }
    SaveModelAction.create = create;
})(SaveModelAction = exports.SaveModelAction || (exports.SaveModelAction = {}));
var SetDirtyStateAction;
(function (SetDirtyStateAction) {
    SetDirtyStateAction.KIND = 'setDirtyState';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, SetDirtyStateAction.KIND) && (0, type_util_1.hasBooleanProp)(object, 'isDirty');
    }
    SetDirtyStateAction.is = is;
    function create(isDirty, options = {}) {
        return Object.assign({ kind: SetDirtyStateAction.KIND, isDirty }, options);
    }
    SetDirtyStateAction.create = create;
})(SetDirtyStateAction = exports.SetDirtyStateAction || (exports.SetDirtyStateAction = {}));
var RequestExportSvgAction;
(function (RequestExportSvgAction) {
    RequestExportSvgAction.KIND = 'requestExportSvg';
    function is(object) {
        return base_protocol_1.RequestAction.hasKind(object, RequestExportSvgAction.KIND);
    }
    RequestExportSvgAction.is = is;
    function create(options = {}) {
        return Object.assign({ kind: RequestExportSvgAction.KIND, requestId: '' }, options);
    }
    RequestExportSvgAction.create = create;
})(RequestExportSvgAction = exports.RequestExportSvgAction || (exports.RequestExportSvgAction = {}));
var ExportSvgAction;
(function (ExportSvgAction) {
    ExportSvgAction.KIND = 'exportSvg';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, ExportSvgAction.KIND) && (0, type_util_1.hasStringProp)(object, 'svg');
    }
    ExportSvgAction.is = is;
    function create(svg, options = {}) {
        return Object.assign({ kind: ExportSvgAction.KIND, svg, responseId: '' }, options);
    }
    ExportSvgAction.create = create;
})(ExportSvgAction = exports.ExportSvgAction || (exports.ExportSvgAction = {}));
