// @eclipse-glsp/protocol/lib/action-protocol/model-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelSourceChangedAction = exports.UpdateModelAction = exports.SetModelAction = exports.RequestModelAction = void 0;
const type_util_1 = require("../utils/type-util");
const base_protocol_1 = require("./base-protocol");
var RequestModelAction;
(function (RequestModelAction) {
    RequestModelAction.KIND = 'requestModel';
    function is(object) {
        return base_protocol_1.RequestAction.hasKind(object, RequestModelAction.KIND);
    }
    RequestModelAction.is = is;
    function create(options = {}) {
        return Object.assign({ kind: RequestModelAction.KIND, requestId: '' }, options);
    }
    RequestModelAction.create = create;
})(RequestModelAction = exports.RequestModelAction || (exports.RequestModelAction = {}));
var SetModelAction;
(function (SetModelAction) {
    SetModelAction.KIND = 'setModel';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, SetModelAction.KIND) && (0, type_util_1.hasObjectProp)(object, 'newRoot');
    }
    SetModelAction.is = is;
    function create(newRoot, options = {}) {
        return Object.assign({ kind: SetModelAction.KIND, responseId: '', newRoot }, options);
    }
    SetModelAction.create = create;
})(SetModelAction = exports.SetModelAction || (exports.SetModelAction = {}));
var UpdateModelAction;
(function (UpdateModelAction) {
    UpdateModelAction.KIND = 'updateModel';
    function is(action) {
        return base_protocol_1.Action.hasKind(action, UpdateModelAction.KIND) && (0, type_util_1.hasObjectProp)(action, 'newRoot');
    }
    UpdateModelAction.is = is;
    function create(newRoot, options = {}) {
        return Object.assign({ kind: UpdateModelAction.KIND, newRoot, animate: true }, options);
    }
    UpdateModelAction.create = create;
})(UpdateModelAction = exports.UpdateModelAction || (exports.UpdateModelAction = {}));
var ModelSourceChangedAction;
(function (ModelSourceChangedAction) {
    ModelSourceChangedAction.KIND = 'modelSourceChanged';
    function is(object) {
        return base_protocol_1.Action.hasKind(object, ModelSourceChangedAction.KIND) && (0, type_util_1.hasStringProp)(object, 'modelSourceName');
    }
    ModelSourceChangedAction.is = is;
    function create(modelSourceName) {
        return {
            kind: ModelSourceChangedAction.KIND,
            modelSourceName
        };
    }
    ModelSourceChangedAction.create = create;
})(ModelSourceChangedAction = exports.ModelSourceChangedAction || (exports.ModelSourceChangedAction = {}));
