// @eclipse-glsp/protocol/lib/action-protocol/undo-redo.js
"use strict";
/********************************************************************************
 * Copyright (c) 2021-2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedoOperation = exports.UndoOperation = void 0;
const base_protocol_1 = require("./base-protocol");
var UndoOperation;
(function (UndoOperation) {
    UndoOperation.KIND = 'glspUndo';
    function is(object) {
        return base_protocol_1.Operation.hasKind(object, UndoOperation.KIND);
    }
    UndoOperation.is = is;
    function create() {
        return {
            kind: UndoOperation.KIND,
            isOperation: true
        };
    }
    UndoOperation.create = create;
})(UndoOperation = exports.UndoOperation || (exports.UndoOperation = {}));
var RedoOperation;
(function (RedoOperation) {
    RedoOperation.KIND = 'glspRedo';
    function is(object) {
        return base_protocol_1.Operation.hasKind(object, RedoOperation.KIND);
    }
    RedoOperation.is = is;
    function create() {
        return {
            kind: RedoOperation.KIND,
            isOperation: true
        };
    }
    RedoOperation.create = create;
})(RedoOperation = exports.RedoOperation || (exports.RedoOperation = {}));
