// @eclipse-glsp/protocol/lib/action-protocol/viewport.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2022 STMicroelectronics and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
/* eslint-disable max-len */
const chai_1 = require("chai");
const viewport_1 = require("./viewport");
describe('Viewport Actions', () => {
    describe('CenterAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'center',
                    elementIds: [],
                    animate: true,
                    retainZoom: true
                };
                (0, chai_1.expect)(viewport_1.CenterAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(viewport_1.CenterAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(viewport_1.CenterAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'center',
                    elementIds: ['my', 'elements'],
                    animate: true,
                    retainZoom: false
                };
                const { elementIds } = expected;
                (0, chai_1.expect)(viewport_1.CenterAction.create(elementIds)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'center',
                    elementIds: ['my', 'elements'],
                    animate: false,
                    retainZoom: true
                };
                const { elementIds, animate, retainZoom } = expected;
                (0, chai_1.expect)(viewport_1.CenterAction.create(elementIds, { animate, retainZoom })).to.deep.equals(expected);
            });
        });
    });
    describe('FitToScreenAction', () => {
        describe('is', () => {
            it('should return true for an object having the correct type and a value for all required interface properties', () => {
                const action = {
                    kind: 'fit',
                    elementIds: [],
                    animate: true
                };
                (0, chai_1.expect)(viewport_1.FitToScreenAction.is(action)).to.be.true;
            });
            it('should return false for `undefined`', () => {
                (0, chai_1.expect)(viewport_1.FitToScreenAction.is(undefined)).to.be.false;
            });
            it('should return false for an object that does not have all required interface properties', () => {
                (0, chai_1.expect)(viewport_1.FitToScreenAction.is({ kind: 'notTheRightOne' })).to.be.false;
            });
        });
        describe('create', () => {
            it('should return an object conforming to the interface with matching properties for the given required arguments and default values for the optional arguments', () => {
                const expected = {
                    kind: 'fit',
                    elementIds: ['my', 'elements'],
                    animate: true
                };
                const { elementIds } = expected;
                (0, chai_1.expect)(viewport_1.FitToScreenAction.create(elementIds)).to.deep.equals(expected);
            });
            it('should return an object conforming to the interface with matching properties for the given required and optional arguments', () => {
                const expected = {
                    kind: 'fit',
                    elementIds: ['my', 'elements'],
                    animate: false,
                    maxZoom: 4,
                    padding: 12
                };
                const { elementIds, maxZoom, padding, animate } = expected;
                (0, chai_1.expect)(viewport_1.FitToScreenAction.create(elementIds, { maxZoom, padding, animate })).to.deep.equals(expected);
            });
        });
    });
});
