// @eclipse-glsp/client/lib/features/select/select-feedback-action.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectFeedbackCommand = exports.SelectFeedbackAction = void 0;
/********************************************************************************
 * Copyright (c) 2019-2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const protocol_1 = require("@eclipse-glsp/protocol");
const inversify_1 = require("inversify");
const sprotty_1 = require("sprotty");
const types_1 = require("../../base/types");
var SelectFeedbackAction;
(function (SelectFeedbackAction) {
    SelectFeedbackAction.KIND = 'selectFeedback';
    function is(object) {
        return protocol_1.Action.hasKind(object, SelectFeedbackAction.KIND) && (0, protocol_1.hasArrayProp)(object, 'selectedElementsIDs') && (0, protocol_1.hasArrayProp)(object, 'deselectedElementsIDs');
    }
    SelectFeedbackAction.is = is;
    function create(options) {
        return Object.assign(Object.assign({}, protocol_1.SelectAction.create(options)), { kind: SelectFeedbackAction.KIND });
    }
    SelectFeedbackAction.create = create;
})(SelectFeedbackAction = exports.SelectFeedbackAction || (exports.SelectFeedbackAction = {}));
let SelectFeedbackCommand = class SelectFeedbackCommand extends sprotty_1.Command {
    constructor(action) {
        super();
        this.action = action;
        this.sprottySelectCommand = new sprotty_1.SelectCommand(Object.assign(Object.assign({}, action), { kind: protocol_1.SelectAction.KIND }));
    }
    execute(context) {
        return this.sprottySelectCommand.execute(context);
    }
    undo(context) {
        return this.sprottySelectCommand.undo(context);
    }
    redo(context) {
        return this.sprottySelectCommand.redo(context);
    }
};
SelectFeedbackCommand.KIND = SelectFeedbackAction.KIND;
SelectFeedbackCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __metadata("design:paramtypes", [Object])
], SelectFeedbackCommand);
exports.SelectFeedbackCommand = SelectFeedbackCommand;
