// @eclipse-glsp/client/lib/features/select/selection-service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectAllCommand = exports.SelectCommand = exports.SelectionService = void 0;
/********************************************************************************
 * Copyright (c) 2019-2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const protocol_1 = require("@eclipse-glsp/protocol");
const inversify_1 = require("inversify");
const sprotty_1 = require("sprotty");
const types_1 = require("../../base/types");
const smodel_util_1 = require("../../utils/smodel-util");
const select_feedback_action_1 = require("./select-feedback-action");
let SelectionService = class SelectionService {
    constructor(selectionListeners = []) {
        this.selectionListeners = selectionListeners;
        this.selectedElementIDs = new Set();
    }
    register(selectionListener) {
        (0, protocol_1.distinctAdd)(this.selectionListeners, selectionListener);
    }
    deregister(selectionListener) {
        (0, protocol_1.remove)(this.selectionListeners, selectionListener);
    }
    modelRootChanged(root) {
        this.updateSelection(root, [], []);
    }
    updateSelection(root, select, deselect) {
        if (root === undefined && select.length === 0 && deselect.length === 0) {
            return;
        }
        const prevRoot = this.root;
        const prevSelectedElementIDs = new Set(this.selectedElementIDs);
        // update root
        this.root = root;
        // update selected element IDs and collect deselected elements
        // - select all elements that are not deselected at the same time (no-op)
        // - deselect all elements that are not selected at the same time (no-op) but was selected
        const toSelect = [...select].filter(selectId => deselect.indexOf(selectId) === -1);
        const toDeselect = [...deselect].filter(deselectId => select.indexOf(deselectId) === -1 && this.selectedElementIDs.has(deselectId));
        for (const id of toDeselect) {
            this.selectedElementIDs.delete(id);
        }
        for (const id of toSelect) {
            this.selectedElementIDs.add(id);
        }
        const deselectedElementIDs = new Set(toDeselect);
        // see if selected elements still exist in the updated root
        for (const id of this.selectedElementIDs) {
            const element = root.index.getById(id);
            if (element === undefined) {
                this.selectedElementIDs.delete(id);
                if (prevRoot !== undefined && prevRoot.index.getById(id)) {
                    deselectedElementIDs.add(id);
                }
            }
        }
        // only send out changes if there actually are changes, i.e., the root or the selected elements changed
        const selectionChanged = prevSelectedElementIDs.size !== this.selectedElementIDs.size ||
            ![...prevSelectedElementIDs].every(value => this.selectedElementIDs.has(value));
        if (selectionChanged) {
            // aggregate to feedback action handling all elements as only the last feedback is restored
            this.dispatchFeedback([
                select_feedback_action_1.SelectFeedbackAction.create({
                    selectedElementsIDs: [...this.selectedElementIDs],
                    deselectedElementsIDs: [...deselectedElementIDs]
                })
            ]);
        }
        const rootChanged = prevRoot !== root;
        if (rootChanged || selectionChanged) {
            // notify listeners after the feedback action
            this.notifyListeners(this.root, this.selectedElementIDs);
        }
    }
    dispatchFeedback(actions) {
        this.feedbackDispatcher.registerFeedback(this, actions);
    }
    notifyListeners(root, selectedElementIDs) {
        this.selectionListeners.forEach(listener => listener.selectionChanged(root, Array.from(selectedElementIDs)));
    }
    getModelRoot() {
        return this.root;
    }
    getSelectedElements() {
        return (0, smodel_util_1.getElements)(this.root.index, Array.from(this.selectedElementIDs), sprotty_1.isSelectable);
    }
    /**
     * QUERY METHODS
     */
    getSelectedElementIDs() {
        return [...this.selectedElementIDs];
    }
    hasSelectedElements() {
        return this.selectedElementIDs.size > 0;
    }
    isSingleSelection() {
        return this.selectedElementIDs.size === 1;
    }
    isMultiSelection() {
        return this.selectedElementIDs.size > 1;
    }
};
__decorate([
    (0, inversify_1.inject)(types_1.TYPES.IFeedbackActionDispatcher),
    __metadata("design:type", Object)
], SelectionService.prototype, "feedbackDispatcher", void 0);
__decorate([
    (0, inversify_1.inject)(types_1.TYPES.ILogger),
    __metadata("design:type", Object)
], SelectionService.prototype, "logger", void 0);
SelectionService = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.multiInject)(types_1.TYPES.SelectionListener)),
    __param(0, (0, inversify_1.optional)()),
    __metadata("design:paramtypes", [Array])
], SelectionService);
exports.SelectionService = SelectionService;
/**
 * Handles a {@link SelectAction} and propagates the new selection to the {@link SelectionService}.
 * Other tools might be selection-sensitive which means {@link SelectAction}s must be processed as fast as possible.
 * Handling the action with a command ensures that the action is processed before the next render tick.
 */
let SelectCommand = class SelectCommand extends sprotty_1.Command {
    constructor(action, selectionService) {
        super();
        this.action = action;
        this.selectionService = selectionService;
        this.selected = [];
        this.deselected = [];
    }
    execute(context) {
        const model = context.root;
        const selectionGuard = (element) => element instanceof sprotty_1.SChildElement && (0, sprotty_1.isSelectable)(element);
        const selectedElements = (0, smodel_util_1.getElements)(model.index, this.action.selectedElementsIDs, selectionGuard);
        const deselectedElements = (0, smodel_util_1.getElements)(model.index, this.action.deselectedElementsIDs, selectionGuard);
        this.selectionService.updateSelection(model, (0, protocol_1.pluck)(selectedElements, 'id'), (0, protocol_1.pluck)(deselectedElements, 'id'));
        return model;
    }
    // Basically no-op since client-side undo is not supported in GLSP.
    undo(context) {
        return context.root;
    }
    // Basically no-op since client-side redo is not supported in GLSP.
    redo(context) {
        return context.root;
    }
};
SelectCommand.KIND = sprotty_1.SelectCommand.KIND;
SelectCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __param(1, (0, inversify_1.inject)(types_1.TYPES.SelectionService)),
    __metadata("design:paramtypes", [Object, SelectionService])
], SelectCommand);
exports.SelectCommand = SelectCommand;
/**
 * Handles a {@link SelectAllAction} and propagates the new selection to the {@link SelectionService}.
 * Other tools might be selection-sensitive which means {@link SelectionAllAction}s must be processed as fast as possible.
 * Handling the action with a command ensures that the action is processed before the next render tick.
 */
let SelectAllCommand = class SelectAllCommand extends sprotty_1.Command {
    constructor(action, selectionService) {
        super();
        this.action = action;
        this.selectionService = selectionService;
        this.previousSelection = new Map();
    }
    execute(context) {
        const model = context.root;
        const selectionGuard = (element) => element instanceof sprotty_1.SChildElement && (0, sprotty_1.isSelectable)(element);
        const selectables = (0, smodel_util_1.getMatchingElements)(model.index, selectionGuard);
        const selectableIds = (0, protocol_1.pluck)(selectables, 'id');
        if (this.action.select) {
            this.selectionService.updateSelection(model, selectableIds, []);
        }
        else {
            this.selectionService.updateSelection(model, [], selectableIds);
        }
        return model;
    }
    // Basically no-op since client-side undo is not supported in GLSP.
    undo(context) {
        return context.root;
    }
    // Basically no-op since client-side redo is not supported in GLSP.
    redo(context) {
        return context.root;
    }
};
SelectAllCommand.KIND = sprotty_1.SelectAllCommand.KIND;
SelectAllCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __param(1, (0, inversify_1.inject)(types_1.TYPES.SelectionService)),
    __metadata("design:paramtypes", [Object, SelectionService])
], SelectAllCommand);
exports.SelectAllCommand = SelectAllCommand;
