// @eclipse-glsp/theia-integration/lib/browser/diagram/glsp-diagram-manager.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveDiagramManagerId = exports.GLSPDiagramManager = exports.TheiaGLSPConnectorProvider = exports.registerDiagramManager = void 0;
/********************************************************************************
 * Copyright (c) 2019-2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const client_1 = require("@eclipse-glsp/client");
const browser_1 = require("@theia/core/lib/browser");
const selection_service_1 = require("@theia/core/lib/common/selection-service");
const inversify_1 = require("@theia/core/shared/inversify");
const browser_2 = require("@theia/editor/lib/browser");
const sprotty_theia_1 = require("sprotty-theia");
const theia_opener_options_navigation_service_1 = require("../theia-opener-options-navigation-service");
const glsp_diagram_context_key_service_1 = require("./glsp-diagram-context-key-service");
const glsp_diagram_widget_1 = require("./glsp-diagram-widget");
function registerDiagramManager(bind, diagramManagerServiceId, bindToSelf = true) {
    if (bindToSelf) {
        bind(diagramManagerServiceId).toSelf().inSingletonScope();
    }
    bind(sprotty_theia_1.DiagramManager).toService(diagramManagerServiceId);
    bind(browser_1.FrontendApplicationContribution).toService(diagramManagerServiceId);
    bind(browser_1.OpenHandler).toService(diagramManagerServiceId);
    bind(browser_1.WidgetFactory).toService(diagramManagerServiceId);
    bind(sprotty_theia_1.DiagramManagerProvider).toProvider(context => () => new Promise(resolve => {
        resolve(context.container.get(diagramManagerServiceId));
    }));
}
exports.registerDiagramManager = registerDiagramManager;
exports.TheiaGLSPConnectorProvider = Symbol('TheiaGLSPConnectorProvider');
let GLSPDiagramManager = class GLSPDiagramManager extends sprotty_theia_1.DiagramManager {
    async initialize() {
        this._diagramConnector = await this.connectorProvider(this.diagramType);
        if (!this._diagramConnector) {
            throw new Error(`No diagram connector is registered for diagramType: ${this.diagramType}!`);
        }
    }
    async doOpen(widget, options) {
        const widgetWasAttached = widget.isAttached;
        await super.doOpen(widget);
        const navigations = this.diagramNavigationService.determineNavigations(widget.uri.toString(true), options);
        if (navigations.length > 0) {
            if (widget.actionDispatcher instanceof client_1.GLSPActionDispatcher) {
                widget.actionDispatcher.onceModelInitialized().then(() => widget.actionDispatcher.dispatchAll(navigations));
            }
            else {
                widget.actionDispatcher.dispatchAll(navigations);
            }
        }
        else if (!widgetWasAttached && widget instanceof glsp_diagram_widget_1.GLSPDiagramWidget) {
            widget.restoreViewportDataFromStorageService();
        }
    }
    async createWidget(options) {
        if (sprotty_theia_1.DiagramWidgetOptions.is(options)) {
            const clientId = this.createClientId();
            const widgetId = this.createWidgetId(options);
            const config = this.getDiagramConfiguration(options);
            const diContainer = config.createContainer(clientId);
            const initializeResult = await this.diagramConnector.initializeResult;
            await (0, client_1.configureServerActions)(initializeResult, this.diagramType, diContainer);
            const widget = new glsp_diagram_widget_1.GLSPDiagramWidget(options, widgetId, diContainer, this.editorPreferences, this.storage, this.theiaSelectionService, this.diagramConnector);
            widget.listenToFocusState(this.shell);
            return widget;
        }
        throw Error('DiagramWidgetFactory needs DiagramWidgetOptions but got ' + JSON.stringify(options));
    }
    createWidgetOptions(uri, options) {
        return {
            diagramType: this.diagramType,
            kind: 'navigatable',
            uri: uri.toString(true),
            iconClass: this.iconClass,
            label: uri.path.base,
            editMode: options && options.editMode ? options.editMode : client_1.EditMode.EDITABLE
        };
    }
    createWidgetId(options) {
        return `${this.diagramType}:${options.uri}`;
    }
    getDiagramConfiguration(options) {
        return this.diagramConfigurationRegistry.get(options.diagramType);
    }
    canHandle(uri, _options) {
        for (const extension of this.fileExtensions) {
            if (uri.path.toString().endsWith(extension)) {
                return 1001;
            }
        }
        return 0;
    }
    get diagramConnector() {
        return this._diagramConnector;
    }
    get id() {
        return deriveDiagramManagerId(this.diagramType);
    }
    get iconClass() {
        return (0, client_1.codiconCSSString)('type-hierarchy-sub');
    }
};
__decorate([
    (0, inversify_1.inject)(browser_2.EditorPreferences),
    __metadata("design:type", Object)
], GLSPDiagramManager.prototype, "editorPreferences", void 0);
__decorate([
    (0, inversify_1.inject)(browser_1.StorageService),
    __metadata("design:type", Object)
], GLSPDiagramManager.prototype, "storage", void 0);
__decorate([
    (0, inversify_1.inject)(theia_opener_options_navigation_service_1.TheiaOpenerOptionsNavigationService),
    __metadata("design:type", theia_opener_options_navigation_service_1.TheiaOpenerOptionsNavigationService)
], GLSPDiagramManager.prototype, "diagramNavigationService", void 0);
__decorate([
    (0, inversify_1.inject)(glsp_diagram_context_key_service_1.GLSPDiagramContextKeyService),
    __metadata("design:type", glsp_diagram_context_key_service_1.GLSPDiagramContextKeyService)
], GLSPDiagramManager.prototype, "contextKeyService", void 0);
__decorate([
    (0, inversify_1.inject)(browser_1.ApplicationShell),
    __metadata("design:type", browser_1.ApplicationShell)
], GLSPDiagramManager.prototype, "shell", void 0);
__decorate([
    (0, inversify_1.inject)(selection_service_1.SelectionService),
    __metadata("design:type", selection_service_1.SelectionService)
], GLSPDiagramManager.prototype, "theiaSelectionService", void 0);
__decorate([
    (0, inversify_1.inject)(exports.TheiaGLSPConnectorProvider),
    __metadata("design:type", Function)
], GLSPDiagramManager.prototype, "connectorProvider", void 0);
__decorate([
    (0, inversify_1.postConstruct)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], GLSPDiagramManager.prototype, "initialize", null);
GLSPDiagramManager = __decorate([
    (0, inversify_1.injectable)()
], GLSPDiagramManager);
exports.GLSPDiagramManager = GLSPDiagramManager;
function deriveDiagramManagerId(diagramType) {
    return diagramType + '-diagram-manager';
}
exports.deriveDiagramManagerId = deriveDiagramManagerId;
