// @eclipse-glsp/client/lib/features/tool-feedback/css-feedback.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteCssClasses = exports.applyCssClasses = exports.cursorFeedbackAction = exports.CursorCSS = exports.ModifyCssFeedbackCommand = exports.ModifyCSSFeedbackAction = void 0;
/********************************************************************************
 * Copyright (c) 2019-2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
const protocol_1 = require("@eclipse-glsp/protocol");
const inversify_1 = require("inversify");
const types_1 = require("../../base/types");
const smodel_util_1 = require("../../utils/smodel-util");
const model_1 = require("./model");
var ModifyCSSFeedbackAction;
(function (ModifyCSSFeedbackAction) {
    ModifyCSSFeedbackAction.KIND = 'modifyCSSFeedback';
    function is(object) {
        return protocol_1.Action.hasKind(object, ModifyCSSFeedbackAction.KIND) && (0, protocol_1.hasArrayProp)(object, 'elementIds');
    }
    ModifyCSSFeedbackAction.is = is;
    function create(options) {
        const { elements } = options, rest = __rest(options, ["elements"]);
        const elementIds = elements ? elements.map(element => (typeof element === 'string' ? element : element.id)) : undefined;
        return Object.assign({ kind: ModifyCSSFeedbackAction.KIND, elementIds }, rest);
    }
    ModifyCSSFeedbackAction.create = create;
})(ModifyCSSFeedbackAction = exports.ModifyCSSFeedbackAction || (exports.ModifyCSSFeedbackAction = {}));
let ModifyCssFeedbackCommand = class ModifyCssFeedbackCommand extends model_1.FeedbackCommand {
    constructor(action) {
        super();
        this.action = action;
    }
    execute(context) {
        const elements = this.action.elementIds ? (0, smodel_util_1.getElements)(context.root.index, this.action.elementIds) : [context.root];
        elements.forEach(e => {
            if (this.action.remove) {
                (0, smodel_util_1.removeCssClasses)(e, this.action.remove);
            }
            if (this.action.add) {
                (0, smodel_util_1.addCssClasses)(e, this.action.add);
            }
        });
        return context.root;
    }
};
ModifyCssFeedbackCommand.KIND = ModifyCSSFeedbackAction.KIND;
ModifyCssFeedbackCommand = __decorate([
    (0, inversify_1.injectable)(),
    __param(0, (0, inversify_1.inject)(types_1.TYPES.Action)),
    __metadata("design:paramtypes", [Object])
], ModifyCssFeedbackCommand);
exports.ModifyCssFeedbackCommand = ModifyCssFeedbackCommand;
// eslint-disable-next-line no-shadow
var CursorCSS;
(function (CursorCSS) {
    CursorCSS["DEFAULT"] = "default-mode";
    CursorCSS["OVERLAP_FORBIDDEN"] = "overlap-forbidden-mode";
    CursorCSS["NODE_CREATION"] = "node-creation-mode";
    CursorCSS["EDGE_CREATION_SOURCE"] = "edge-creation-select-source-mode";
    CursorCSS["EDGE_CREATION_TARGET"] = "edge-creation-select-target-mode";
    CursorCSS["EDGE_RECONNECT"] = "edge-reconnect-select-target-mode";
    CursorCSS["OPERATION_NOT_ALLOWED"] = "edge-modification-not-allowed-mode";
    CursorCSS["ELEMENT_DELETION"] = "element-deletion-mode";
    CursorCSS["RESIZE_NESW"] = "resize-nesw-mode";
    CursorCSS["RESIZE_NWSE"] = "resize-nwse-mode";
    CursorCSS["MOVE"] = "move-mode";
    CursorCSS["MARQUEE"] = "marquee-mode";
})(CursorCSS = exports.CursorCSS || (exports.CursorCSS = {}));
function cursorFeedbackAction(cursorCss) {
    const add = [];
    if (cursorCss) {
        add.push(cursorCss);
    }
    return ModifyCSSFeedbackAction.create({ add, remove: Object.values(CursorCSS) });
}
exports.cursorFeedbackAction = cursorFeedbackAction;
function applyCssClasses(element, ...add) {
    return ModifyCSSFeedbackAction.create({ elements: [element], add });
}
exports.applyCssClasses = applyCssClasses;
function deleteCssClasses(element, ...remove) {
    return ModifyCSSFeedbackAction.create({ elements: [element], remove });
}
exports.deleteCssClasses = deleteCssClasses;
